plugins/com.google.cloud.tools.eclipse.util/src/com/google/cloud/tools/eclipse/util/MappedNamespaceContext.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.eclipse.util;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import javax.xml.XMLConstants;
+import javax.xml.namespace.NamespaceContext;
+
+public class MappedNamespaceContext implements NamespaceContext {
+
+  private final Map<String, String> prefixMapping = new HashMap<>();
+
+  public MappedNamespaceContext() {
+    prefixMapping.put("xml", XMLConstants.XML_NS_URI);
+  }
+
+  public MappedNamespaceContext(String prefix, String namespaceUri) {
+    this();
+    addMappingInternal(prefix, namespaceUri);
+  }
+
+  public MappedNamespaceContext addMapping(String prefix, String namespaceUri) {
+    return addMappingInternal(prefix, namespaceUri);
+  }
+
+  private MappedNamespaceContext addMappingInternal(String prefix, String namespaceUri) {
```

Comment:
One more special case to handle:

The prefix xmlns is used only to declare namespace bindings and is by definition bound to the namespace name http://www.w3.org/2000/xmlns/. It MUST NOT be declared . Other prefixes MUST NOT be bound to this namespace name, and it MUST NOT be declared as the default namespace. Element names MUST NOT have the prefix xmlns.

except

The prefix xmlns is used only for namespace bindings and is not itself bound to any namespace name.

That is, this rule changed between editions of the spec. I think we should throw an IllegalArgumentException if xmlns is passed to this method as a prefix.